### ISSUE TITLE
Reads from namespace buckets fail 

### ISSUE THREAD

**Issue by gbadanahatti** [noobaa/noobaa-core#6008]
### Environment info
- NooBaa Version:
INFO[0000] CLI version: 2.1.0                           
INFO[0000] noobaa-image: noobaa/noobaa-core:5.3.0       
INFO[0000] operator-image: noobaa/noobaa-operator:2.1.0 

- Platform: **Kubernetes 1.14.1 | minikube 1.1.1 | OpenShift 4.1 | other: specify**
Kubernetes 1.14.1

### Actual behavior
1. Reads from namespace buckets fail with integrity checks.

com.amazonaws.AmazonClientException: Unable to verify integrity of data download.  Client calculated content hash didn't match hash calculated by Amazon S3.  The data may be corrupt.
        at com.amazonaws.services.s3.internal.DigestValidationInputStream.validateMD5Digest(DigestValidationInputStream.java:79)### Expected behavior

2. Also, it looks like Noobaa is performing range reads of 4K only from Object Storage.

### Steps to reproduce
1. Create a Namespace resource and a namespace bucket.
2. Write to the namespace resource through the namespace bucket.
3. Read back the written objects.

**Comment by jeniawhite** [comment]
@gbadanahatti 
Hello, sorry for the late response.
I've attempted to reproduce on Minikube and AWS as target namespace using the s3cli.
Everything worked like a charm.
```
(base) jeniawhite@MacBook-Pro noobaa-operator % aws s3api --endpoint-url https://192.168.64.37:30483 --no-verify-ssl put-object --bucket jenia --key jenia.txt --body ~/Downloads/noobaa-core-0.logs 
/usr/local/Cellar/awscli/2.0.0/libexec/lib/python3.8/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '192.168.64.37'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "ETag": "\"1349597d125239b6f77cc44cd1f1ca3c\""
}
(base) jeniawhite@MacBook-Pro noobaa-operator % aws s3api --endpoint-url https://192.168.64.37:30483 --no-verify-ssl get-object --bucket jenia --key jenia.txt ~/Desktop/noobaa-core-0.logs
/usr/local/Cellar/awscli/2.0.0/libexec/lib/python3.8/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '192.168.64.37'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "AcceptRanges": "bytes",
    "LastModified": "2020-05-14T18:43:15+00:00",
    "ContentLength": 21953226,
    "ETag": "\"1349597d125239b6f77cc44cd1f1ca3c\"",
    "ContentType": "text/plain",
    "Metadata": {
        "noobaa-namespace-s3-bucket": "jenia"
    }
}
```
After downloading I've checked the md5 and receieved exactly the same file.
```
(base) jeniawhite@MacBook-Pro noobaa-operator % md5 ~/Downloads/noobaa-core-0.logs
MD5 (/Users/jeniawhite/Downloads/noobaa-core-0.logs) = 1349597d125239b6f77cc44cd1f1ca3c
(base) jeniawhite@MacBook-Pro noobaa-operator % md5 ~/Desktop/noobaa-core-0.logs 
MD5 (/Users/jeniawhite/Desktop/noobaa-core-0.logs) = 1349597d125239b6f77cc44cd1f1ca3c
```
Could you please elaborate more on how I shall reproduce your problem?
Thank you.

**Comment by gbadanahatti** [comment]
@jeniawhite: I am using IBM COS backend. Would that make a difference?
I also noticed that the range reads of the objects are only 4K in size. I'll send the endpoint logs.

**Comment by jeniawhite** [comment]
@gbadanahatti 
It would be great if you could send the exact code that you are running with the scenario so I could reproduce and investigate.
I would also need access to IBM COS backend to reproduce (should be using the same namespace but just in case).

**Comment by gbadanahatti** [comment]
@jeniawhite 
I am using cosbench to read the objects back. I have put the cosbench file, cosbench logs and endpoint log in the tar file attachment.

Also, reading the object from awscli results only in the 4K read while the object is actually 64K

Read from noobaa ns bucket:

[root@ait-cosbench-master-1 ~]# aws s3api get-object  --profile noobaa --endpoint http://172.20.29.124:30801 --bucket noobaa-ns-1 --key 64K_100 /tmp/64K_1
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Wed, 06 May 2020 14:40:43 GMT", 
    "ContentLength": **4096**, 
    "ETag": "\"a21195dbd1400e70ecf58ecc151afa5a\"", 
    "Metadata": {
        "noobaa-namespace-s3-bucket": "noobaa-ns-1"
    }
}

Read from IBM COS backend:

 aws s3api get-object  --profile dsnet --endpoint http://172.20.29.11 --bucket noobaa-ns-1 --key 64K_100 /tmp/64_1_1
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Wed, 06 May 2020 14:40:43 GMT", 
    "ContentLength": **64000**, 
    "ETag": "\"a21195dbd1400e70ecf58ecc151afa5a\"", 
    "Metadata": {}
}

[logs.tar.gz](https://github.com/noobaa/noobaa-core/files/4644777/logs.tar.gz)

**Comment by jeniawhite** [comment]
@gbadanahatti 
I will attempt to get an IBM COS endpoint to reproduce the issue.
Will update on my findings, thank you.

**Comment by jeniawhite** [comment]
@gbadanahatti 
Used IBM COS as the endpoint for the Namespace bucket.
Uploaded a file with size 56k using a put-object operation with s3api.
Downloaded it from the IBM COS endpoint and from NooBaa endpoint.
You can see that I've read the same file in all cases and compared the hashes.
```
(base) jeniawhite@MacBook-Pro noobaa-core-code % aws s3api --endpoint-url https://192.168.64.37:30483 put-object --bucket jns --key jen --body ~/Downloads/install.zip --no-verify-ssl
/usr/local/Cellar/awscli/2.0.0/libexec/lib/python3.8/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '192.168.64.37'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "ETag": "\"8c441c8bf23c3b40da9dcd694a144238\""
}
(base) jeniawhite@MacBook-Pro noobaa-core-code % aws s3api --endpoint-url https://192.168.64.37:30483 get-object --bucket jns --key jen ~/Desktop/install.zip --no-verify-ssl
/usr/local/Cellar/awscli/2.0.0/libexec/lib/python3.8/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host '192.168.64.37'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "AcceptRanges": "bytes",
    "LastModified": "2020-05-20T17:47:23+00:00",
    "ContentLength": 56601,
    "ETag": "\"8c441c8bf23c3b40da9dcd694a144238\"",
    "ContentType": "application/octet-stream",
    "Metadata": {
        "noobaa-namespace-s3-bucket": "jeniabucket"
    }
}
(base) jeniawhite@MacBook-Pro noobaa-core-code % aws s3api --endpoint-url https://s3.fra.eu.cloud-object-storage.appdomain.cloud get-object --bucket jeniabucket --key jen ~/Desktop/COSinstall.zip             
{
    "AcceptRanges": "bytes",
    "LastModified": "2020-05-20T17:47:23+00:00",
    "ContentLength": 56601,
    "ETag": "\"8c441c8bf23c3b40da9dcd694a144238\"",
    "ContentType": "application/octet-stream",
    "Metadata": {}
}
(base) jeniawhite@MacBook-Pro noobaa-core-code % md5 ~/Desktop/install.zip; md5 ~/Downloads/install.zip; md5 ~/Desktop/COSinstall.zip                                                              
MD5 (/Users/jeniawhite/Desktop/install.zip) = 8c441c8bf23c3b40da9dcd694a144238
MD5 (/Users/jeniawhite/Downloads/install.zip) = 8c441c8bf23c3b40da9dcd694a144238
MD5 (/Users/jeniawhite/Desktop/COSinstall.zip) = 8c441c8bf23c3b40da9dcd694a144238
```
I did not install cosbench and just used AWS CLI for uploads and reads, since in your example you've said that even reads with the AWS CLI give you the partial object.
Could you please provide more information on how did you upload the file, or maybe an exact scenario that I can reproduce from A-Z in order to get the same results?

**Comment by gbadanahatti** [comment]
@jeniawhite 
Strange.. I have attached endpoint logs that contain both uploads and downloads. The files were uploaded with cosbench. The tar file contains cosbench configs that uploads the objects and downloads the objects. 

I can see from the COS logs that the upload  is the full 64k and I see that the reads are just the metadata.

Maybe a short webex at a mutually convenient time will help.


[logs_1.tar.gz](https://github.com/noobaa/noobaa-core/files/4662907/logs_1.tar.gz)

**Comment by guymguym** [comment]
@gbadanahatti @jeniawhite Please check if this might be related to this fix by @jackyalbo - https://github.com/noobaa/noobaa-core/pull/5963

Before the fix since we introduced the optimization to use a GET RANGE 0-4K instead of just HEAD so that small objects would be inlined, but we missed returning the correct object size from the ranges. If you are running a version before this fix it would truncate your object reads....

**Comment by jeniawhite** [comment]
@guymguym @gbadanahatti 
Applied a custom image (non-released) with the #5963 fix.
It solved the problems on reads and everything works.
